File: sonar-text-dotnet/src/SonarLint.Secrets.DotNet/Rules/Matching/Match.cs
/*
 * SonarAnalyzer for Text
 * Copyright (C) 2021-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

// NOTE: the Match API in .NET is not the same as the Java version

namespace SonarLint.Secrets.DotNet.Rules.Matching
{
    internal class Match
    {
        public Match(string text, int startIndex, int length)
        {
            Text = text;
            StartIndex = startIndex;
            Length = length;
        }

        public string Text { get; }
        public int StartIndex{ get; }
        public int Length { get; }

        public bool Overlaps(Match other)
        {
            int thisEndIndex = StartIndex + Length - 1;
            int otherEndIndex = other.StartIndex + other.Length - 1;

            return
                // 1. Our start index is inside the other Match
                (StartIndex >= other.StartIndex && StartIndex <= otherEndIndex) ||

                // 2. Our end index is inside the other Match
                (thisEndIndex >= other.StartIndex && thisEndIndex <= otherEndIndex) ||

                // 3. We completely contain the other match
                (this.StartIndex < other.StartIndex && thisEndIndex > otherEndIndex);
        }
    }
}

File: sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/Match.java
/*
 * SonarQube Text Plugin
 * Copyright (C) 2021-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
package org.sonar.plugins.secrets.api;

import java.util.Map;

/**
 * Result of applying pattern matcher to a text.
 *
 * @param text            matched text
 * @param fileStartOffset inclusive start offset
 * @param fileEndOffset   exclusive end offset
 * @param groups          map containing matches per named group
 */
public record Match(String text, int fileStartOffset, int fileEndOffset, Map<String, Match> groups) {

  @Override
  public String toString() {
    return "Match{" +
      "text='" + text + '\'' +
      ", fileStartOffset=" + fileStartOffset +
      ", fileEndOffset=" + fileEndOffset +
      '}';
  }
}
